liblangutil/SourceLocation.h

```diff
@@ -37,28 +38,28 @@ namespace langutil
  */
 struct SourceLocation
 {
-	SourceLocation(): start(-1), end(-1) { }
-	SourceLocation(int _start, int _end, std::shared_ptr<std::string const> _sourceName):
-		start(_start), end(_end), sourceName(_sourceName) { }
+	SourceLocation(): start(-1), end(-1), source{nullptr} { }
+	SourceLocation(int _start, int _end, std::shared_ptr<CharStream> _source):
+		start(_start), end(_end), source{std::move(_source)} { }
 	SourceLocation(SourceLocation&& _other) noexcept:
 		start(_other.start),
 		end(_other.end),
-		sourceName(std::move(_other.sourceName))
+		source{std::move(_other.source)}
 	{}
 	SourceLocation(SourceLocation const&) = default;
 	SourceLocation& operator=(SourceLocation const&) = default;
 	SourceLocation& operator=(SourceLocation&& _other) noexcept
 	{
 		start = _other.start;
 		end = _other.end;
-		sourceName = std::move(_other.sourceName);
+		source = std::move(_other.source);
 		return *this;
 	}
 
 	bool operator==(SourceLocation const& _other) const
 	{
 		return start == _other.start && end == _other.end &&
-			((!sourceName && !_other.sourceName) || (sourceName && _other.sourceName && *sourceName == *_other.sourceName));
+			((!source && !_other.source) || (source && _other.source && source->name() == _other.source->name()));
```

Comment:
I would prefer `source.get() == _other.source.get()` - it would be more accurate with regards to the `--CODEGEN--`-sources. Then of course this whole expression can be simplified.

Reply:
okay, thx.